distribution/lib/Standard/Database/0.0.0-dev/src/DB_Column.enso

```diff
@@ -193,6 +193,107 @@ type DB_Column
+       "reference" `DB_Column` which refers to that unique name. This
+       "reference" column is passed to the callback, which can compute values
+       based on the reference and any other values. Finally, the return value of
+       the callback is wrapped in the binding `WITH ... AS` clause, which is
+       returned from the original call to `with`.
+
+       `with` is only available in backends that support `WITH` clauses inside
+       expressions. For backends that only support a single `WITH` claus at the
+       top level, `with` returns `self` unchanged.
+
+       Semantically, the following expressions will always have the same value:
+
+           1. f column
+           2. column.with f
+
+       ? When to use `with`
+
+         `with` can make queries shorter and/or simpler by eliminating
+         duplicates. However, the `WITH` clause itself, including the bound CTE
+         table name, also takes up space, so if the `self` argument to `with`
+         isn't very large, `with` can actually make the query longer.
+
+         For this reason, it is generally better to apply `with` to values
+         before passing them into a library method, rather than to apply it to
+         the value inside the library method itself.
+
+       > Example
+         Remove duplicates of a large column expression from a query.
+
+            ## Without CTEs
+
+             column = table_builder [['x', [1.3, 4.7, -1.3, -4.7]]] . at "x"
+             rounded = column.round
+             large = rounded * rounded
+             large.to_sql
+             ## =>
+
+                -- Two copies of the complex rounding query
+                SELECT ... [complex rounding query] * [complex rounding query]
+                    ... FROM ...
+
+            ## With CTEs
+
+             not_so_large = column.round.with rounded->
+                rounded * rounded
+             not_so_large.to_sql
+             ## =>
+
+                -- One copy of the complex rounding query
+                SELECT ... (WITH temp_table as ([complex rounding query])
+                                 temp_table.x * temp_table.x)
+                    ... FROM ...
+
+       > Example
+         Use multiple CTEs in a query.
+
+         (column_a * column_b).with product_a_b->
+            (product_a_b * 10).with times_ten->
+                times_ten + product_a_b + 100
+
+       > Example
+         Give names to the CTE table names.
+
+         (column_a * column_b).with name="product_a_b" product_a_b->
+            (product_a_b * 10).with name="times_ten" times_ten->
+                times_ten + product_a_b + 100
+    with : (DB_Column -> DB_Column) -> Text -> DB_Column
+    with self (callback : (DB_Column -> DB_Column)) (name : Text = "with") -> DB_Column =
+        use_ctes = self.connection.dialect.supports_nested_with_clause
+
+        if use_ctes.not then callback self else
+            # TODO: use shorter unique names for binders.
```

Comment:
This uses full UUID table names, which makes the queries longer than they would be if we used shorter unique names. Since the names are temporary and scoped to the query, we cannot query the database to check for name clashes. Using a sequence generator would complicate the implementation (requiring that the sequence state be kept in the connection, or using renaming during SQL generation in `Base_Generator`). This implementation is sufficient to make the `round` queries much smaller.

Reply:
> we cannot query the database to check for name clashes.

We don't have to - I don't think it's an error to use an existing table name as an `AS` name in the query - as long as the names are unique within our query we are good.

Reply:
Yes, I checked and you can use a table name as a WITH..AS binding, but there is the (very slight) possibility of shadowing a table that you didn't mean to, for example:

````
create temp table foo (a int);
insert into foo (a) values (10);
insert into foo (a) values (20);

select a*a from foo;

with foo as (select a+1 as a from foo)
    select a*a from foo;
```

```
 ?column?
----------
      100
      400
(2 rows)

 ?column?
----------
      121
      441
(2 rows)
```

Reply:
Yeah I know - I meant that we can clash with tables in the **Database**, but as for names within the query we should indeed be unique - but these are easily discovered at 'compile time'.

Reply:
I think it would be great if we could try getting the shorter temp-variable names for this PR, as the names add a lot of overhead and make the query less readable.

Isn't it enough if names are unique in the specific 'IR subtree' we are in? E.g. we could have a function that lists all 'taken' names for a given IR Context and IR Expression and generate fresh ones in regard to that. Or would there still be meaningful clashes between 'neighboring' expressions with this strategy?

Reply:
Ah, I see, you were talking about clashes within the query only.I think there would only be a problem if a with-binding name were used twice, and were both in scope in one part of the IR tree. So the best thing to do would be to simply make them unique, like you say.

What I can do right now is remap the unique names to shorter ones during query generation in Base_Generator. (Doing it in `DB_Column.with` would require some kind perstent state.) I avoided doing that because it would make it a bit more complicated, but it's really not that bad and would really help make the SQL shorter. It's a good idea to just scan the entire IR for table names so I can guarantee there are no duplicates.

Reply:
It would be great if you could try this. This would be a first IR optimization pass that we'll have :)

Reply:
Done